sql/planner/tree.go

```diff
@@ -0,0 +1,509 @@
+}
+
+// NewTree creates a new tree with n as root.
+func NewTree(n Node) *Tree {
+	return &Tree{Root: n}
+}
+
+// Run implements the query.Statement interface.
+// It binds the tree to the database resources and executes it.
+func (t *Tree) Run(tx *database.Transaction, params []expr.Param) (query.Result, error) {
+	err := Bind(t, tx, params)
+	if err != nil {
+		return query.Result{}, err
+	}
+
+	t, err = Optimize(t)
+	if err != nil {
+		return query.Result{}, err
+	}
+
+	return t.execute()
+}
+
+func (t *Tree) execute() (query.Result, error) {
+	var st document.Stream
+	var err error
+
+	if t.Root.Left() != nil {
+		st, err = nodeToStream(t.Root.Left())
+		if err != nil {
+			return query.Result{}, err
+		}
+	}
+
+	return t.Root.(outputNode).toResult(st)
+}
+
+// IsReadOnly implements the query.Statement interface.
+func (t *Tree) IsReadOnly() bool {
+	return false
+}
+
+func nodeToStream(n Node) (st document.Stream, err error) {
+	l := n.Left()
+	if l != nil {
+		st, err = nodeToStream(l)
+		if err != nil {
+			return
+		}
+	}
+
+	switch t := n.(type) {
+	case inputNode:
+		st, err = t.buildStream()
+	case operationNode:
+		st, err = t.toStream(st)
+	default:
+		panic(fmt.Sprintf("incorrect node type %#v", n))
+	}
+
+	return
+}
+
+// A Node represents an operation on the stream.
+type Node interface {
+	Operation() Operation
+	Left() Node
+	Right() Node
+	SetLeft(Node)
+	SetRight(Node)
+	Equal(Node) bool
+	Bind(tx *database.Transaction, params []expr.Param) error
+}
+
+type inputNode interface {
+	Node
+
+	buildStream() (document.Stream, error)
+}
+
+type operationNode interface {
+	Node
+
+	toStream(st document.Stream) (document.Stream, error)
+}
+
+type outputNode interface {
+	Node
+
+	toResult(st document.Stream) (query.Result, error)
+}
+
+type node struct {
+	op          Operation
+	left, right Node
+}
+
+func (n *node) Operation() Operation {
+	return n.op
+}
+
+func (n *node) Left() Node {
+	return n.left
+}
+
+func (n *node) Right() Node {
+	return n.right
+}
+
+func (n *node) SetLeft(ln Node) {
+	n.left = ln
+}
+
+func (n *node) SetRight(rn Node) {
+	n.right = rn
+}
+
+func (n *node) Equal(other Node) bool {
+	if n == nil {
+		return other == nil
+	}
+
+	if other == nil {
+		return false
+	}
+
+	if n.op != other.Operation() {
+		return false
+	}
+
+	if n.left == nil {
+		if other.Left() != nil {
+			return false
+		}
+	} else if !n.left.Equal(other.Left()) {
+		return false
+	}
+
+	if n.right == nil {
+		if other.Right() != nil {
+			return false
+		}
+	} else if !n.right.Equal(other.Right()) {
+		return false
+	}
+
+	return true
+}
+
+type selectionNode struct {
+	node
+
+	cond   expr.Expr
+	tx     *database.Transaction
+	params []expr.Param
+}
+
+var _ operationNode = (*selectionNode)(nil)
+
+// NewSelectionNode creates a node that filters documents of a stream, according to
+// the expression condition.
+func NewSelectionNode(n Node, cond expr.Expr) Node {
+	return &selectionNode{
+		node: node{
+			op:   Selection,
+			left: n,
```

Comment:
Is there any specific reason to set the left node?

Reply:
Since we currently only read from one table, we only use one branch of the tree, that's why I only used the left branch of each node